bolt-boost/src/proofs.rs

```diff
@@ -0,0 +1,226 @@
+use alloy::primitives::{TxHash, B256};
+
+use super::types::{ConstraintsWithProofData, InclusionProofs};
+
+#[derive(Debug, thiserror::Error)]
+pub enum ProofError {
+    #[error("Leaves and indices length mismatch")]
+    LengthMismatch,
+    #[error("Mismatch in provided leaves and leaves to prove")]
+    LeavesMismatch,
+    #[error("Hash not found in constraints cache: {0:?}")]
+    MissingHash(TxHash),
+    #[error("Proof verification failed")]
+    VerificationFailed,
+}
+
+/// Returns the length of the leaves that need to be proven (i.e. all transactions).
+fn total_leaves(constraints: &[ConstraintsWithProofData]) -> usize {
+    constraints.iter().map(|c| c.proof_data.len()).sum()
+}
+
+/// Verifies the provided multiproofs against the constraints & transactions root.
+/// TODO: support bundle proof verification a.k.a. relative ordering!
+pub fn verify_multiproofs(
+    constraints: &[ConstraintsWithProofData],
+    proofs: &InclusionProofs,
+    root: B256,
+) -> Result<(), ProofError> {
+    // Check if the length of the leaves and indices match
+    if proofs.transaction_hashes.len() != proofs.generalized_indeces.len() {
+        return Err(ProofError::LengthMismatch);
+    }
+
+    let total_leaves = total_leaves(constraints);
+
+    // Check if the total leaves matches the proofs provided
+    if total_leaves != proofs.total_leaves() {
+        return Err(ProofError::LeavesMismatch);
+    }
+
+    // Get all the leaves from the saved constraints
+    let mut leaves = Vec::with_capacity(proofs.total_leaves());
+
+    for hash in &proofs.transaction_hashes {
+        let mut found = false;
+        for constraint in constraints {
+            for (saved_hash, leaf) in &constraint.proof_data {
+                if saved_hash == hash {
+                    found = true;
+                    leaves.push(B256::from(leaf.0));
+                    break;
+                }
+            }
+            if found {
+                break;
+            }
+        }
+
+        // If the hash is not found in the constraints cache, return an error
+        if !found {
+            return Err(ProofError::MissingHash(*hash));
+        }
+    }
+
+    // Verify the Merkle multiproof against the root
+    ssz_rs::multiproofs::verify_merkle_multiproof(
+        &leaves,
+        &proofs.merkle_hashes,
+        &proofs.generalized_indeces,
+        root,
+    )
+    .map_err(|_| ProofError::VerificationFailed)?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use alloy::{
+        hex::FromHex,
+        primitives::{Bytes, B256},
+    };
+    use ssz_rs::{HashTreeRoot, List, PathElement, Prove};
+
+    use crate::testutil::*;
+
+    // #[test]
+    // fn test_single_multiproof() {
+    //     let (root, transactions) = read_test_transactions();
```

Comment:
Is this intended to be left as a comment?

Reply:
Added a note! Multiproof generation is not supported in main yet.